File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/intent/ReadIntentCompiler.java
package com.foundgine.core.semantic.intent;

import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.*;
import com.foundgine.core.semantic.ir.*;
import com.foundgine.core.semantic.ir.graph.SemanticOperationGraph;
import com.foundgine.core.semantic.query.*;
import com.foundgine.core.semantic.resolution.SemanticRequestResolver;
import java.util.*;

/** Compiles external structured read intent into canonical provider-neutral semantic IR. */
public final class ReadIntentCompiler {
    private final SemanticModel model;
    private final SemanticContractSnapshot contract;

    public ReadIntentCompiler(SemanticModel model) { this.model=Objects.requireNonNull(model); this.contract=null; }
    public ReadIntentCompiler(SemanticContractSnapshot contract) { this.contract=Objects.requireNonNull(contract); this.model=null; }

    public String contractFingerprint() {
        if (contract != null) return contract.contractFingerprint();
        return model.freeze().contractFingerprint();
    }

    private SemanticContractSnapshot snapshot() {
        if (contract != null) return contract;
        return new SemanticContractSnapshot(model.freeze());
    }

    public SemanticRequest compile(ReadIntent intent) {
        Objects.requireNonNull(intent);
        var c=snapshot();
        var root=findEntity(c,intent.rootEntity());
        var selections=intent.selections().stream().map(s->compileSelection(c,root,s)).toList();
        var filter=intent.filter()==null?null:compileFilter(c,root,intent.filter());
        var order=intent.order().stream().map(o->compileOrder(c,root,o)).toList();
        return new SemanticRequest(root.id(),selections,new SemanticQueryOptions(filter,order,intent.limit(),intent.offset(),intent.after()),intent.security());
    }

    public SemanticOperationGraph compileOperationGraph(ReadIntent intent) {
        var request=compile(intent);
        var graph=new SemanticRequestResolver(snapshot()).resolve(request);
        return SemanticOperationGraph.create(SemanticOperationCompiler.compile(graph));
    }

    public SemanticIntentDocument createDocument(ReadIntent intent) { return new SemanticIntentDocument(contractFingerprint(),intent).validate(); }

    public SemanticIntentResolution resolveDocument(SemanticIntentDocument document) {
        Objects.requireNonNull(document).validate();
        if (!document.contractFingerprint().equals(contractFingerprint()))
            throw new IllegalArgumentException("Semantic intent document is bound to contract '"+document.contractFingerprint()+"', but the resolver uses '"+contractFingerprint()+"'.");
        return new SemanticIntentResolution(document,compile(document.intent()),contractFingerprint());
    }

    public SemanticOperationGraph resolveDocumentGraph(SemanticIntentDocument document) { return compileOperationGraph(resolveDocument(document).document().intent()); }

    private SemanticSelection compileSelection(SemanticContractSnapshot c,SemanticEntity entity,ReadSelection selection) {
        if ((selection.field()==null)==(selection.relationship()==null)) throw invalid("A selection must specify exactly one field or relationship.");
        if (selection.field()!=null) {
            var f=findField(entity,selection.field());
            if (!selection.effectiveChildren().isEmpty()) throw invalid("Field '"+entity.name()+"."+f.name()+"' cannot have children.");
            return new SemanticSelection(f.id(),null,List.of());
        }
        if (selection.effectiveChildren().isEmpty()) throw invalid("Relationship '"+entity.name()+"."+selection.relationship()+"' requires child selections.");
        var path=resolveRelationshipPath(c,entity,selection.relationship());
        var children=selection.effectiveChildren().stream().map(x->compileSelection(c,c.get(path.get(path.size()-1).target()),x)).toList();
        for(int i=path.size()-1;i>=0;i--) children=List.of(new SemanticSelection(null,path.get(i).id(),children));
        return children.get(0);
    }

    private SemanticFilterExpression compileFilter(SemanticContractSnapshot c,SemanticEntity entity,ReadFilter filter) {
        if(filter instanceof ReadFieldFilter f) return new SemanticFieldFilter(findField(entity,f.field()).id(),f.operator(),f.value());
        if(filter instanceof ReadRelationshipFilter r) {
            var path=resolveRelationshipPath(c,entity,r.relationship());
            if(path.size()>1 && r.quantifier()!=SemanticRelationshipQuantifier.SOME) throw invalid("Logical traversal filter currently supports only SOME quantification across multi-hop paths.");
            SemanticFilterExpression p=compileFilter(c,c.get(path.get(path.size()-1).target()),r.predicate());
            for(int i=path.size()-1;i>=0;i--) p=new SemanticRelationshipFilter(path.get(i).id(),i==path.size()-1?r.quantifier():SemanticRelationshipQuantifier.SOME,p);
            return p;
        }
        if(filter instanceof ReadAndFilter a) { if(a.expressions().isEmpty()) throw invalid("AND filter cannot be empty."); return new SemanticAndFilter(a.expressions().stream().map(x->compileFilter(c,entity,x)).toList()); }
        if(filter instanceof ReadOrFilter o) { if(o.expressions().isEmpty()) throw invalid("OR filter cannot be empty."); return new SemanticOrFilter(o.expressions().stream().map(x->compileFilter(c,entity,x)).toList()); }
        throw invalid("Unsupported read filter '"+filter.getClass().getSimpleName()+"'.");
    }

    private SemanticOrderTerm compileOrder(SemanticContractSnapshot c,SemanticEntity root,ReadOrder order) {
        var entity=root; var path=new ArrayList<RelationshipId>();
        for(var name:order.effectivePath()) for(var r:resolveRelationshipPath(c,entity,name)){path.add(r.id());entity=c.get(r.target());}
        var field=order.aggregate()==SemanticOrderAggregate.COUNT?entity.identity().fieldId():findField(entity,order.field()).id();
        return new SemanticOrderTerm(field,order.direction(),path,order.aggregate());
    }

    private SemanticEntity findEntity(SemanticContractSnapshot c,String name){ try{return c.resolveEntity(name);}catch(Exception e){throw invalid("Unknown entity '"+name+"'.");} }
    private static SemanticField findField(SemanticEntity e,String name){return e.fields().stream().filter(f->f.name().equalsIgnoreCase(name)||f.effectiveAliases().stream().anyMatch(a->a.name().equalsIgnoreCase(name))).findFirst().orElseGet(()->{if(e.identity().name().equalsIgnoreCase(name))return new SemanticField(e.identity().fieldId(),e.identity().name(),Object.class);throw invalid("Unknown field '"+e.name()+"."+name+"'.");});}
    private List<SemanticRelationship> resolveRelationshipPath(SemanticContractSnapshot c,SemanticEntity entity,String name){
        var direct=entity.relationships().stream().filter(r->r.name().equalsIgnoreCase(name)||r.effectiveAliases().stream().anyMatch(a->a.name().equalsIgnoreCase(name))).findFirst();
        if(direct.isPresent()) return List.of(direct.get());
        var t=c.traversals().stream().filter(x->x.source().equals(entity.id())&&x.name().equalsIgnoreCase(name)).findFirst().orElseThrow(()->invalid("Unknown relationship or semantic traversal '"+entity.name()+"."+name+"'."));
        var current=entity; var result=new ArrayList<SemanticRelationship>();
        for(var id:t.path()){final var currentEntity=current;var r=current.relationships().stream().filter(x->x.id().equals(id)).findFirst().orElseThrow(()->invalid("Traversal '"+name+"' contains relationship '"+id+"' that is not defined on '"+currentEntity.name()+"'."));result.add(r);current=c.get(r.target());}
        return result;
    }
    private static IllegalArgumentException invalid(String message){return new IllegalArgumentException("Invalid read intent: "+message);}
}

File: src/csharp/Foundgine.Core/Semantic/Intent/ReadIntentCompiler.cs
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Query;
using Foundgine.Core.Semantic.IR;
using Foundgine.Core.Semantic.IR.Graph;
using Foundgine.Core.Semantic.Resolution;
using Foundgine.Core.Semantic.Security.Execution;

namespace Foundgine.Core.Semantic.Intent;

/// <summary>
/// Compiles human/agent-friendly structured intent into the canonical
/// SemanticRequest. It performs only semantic-name resolution; authorization,
/// planning, and execution remain later stages.
/// </summary>
public sealed class ReadIntentCompiler
{
    private readonly SemanticModel? _model;
    private readonly SemanticContractSnapshot? _contract;

    public ReadIntentCompiler(SemanticModel model) =>
        _model = model ?? throw new ArgumentNullException(nameof(model));

    public ReadIntentCompiler(SemanticContractSnapshot contract) =>
        _contract = contract ?? throw new ArgumentNullException(nameof(contract));

    /// <summary>
    /// Compiles a dynamic read intent against the frozen runtime contract and
    /// lowers it to the same canonical operation graph consumed by the planner.
    /// </summary>
    public SemanticOperationGraph CompileOperationGraph(ReadIntent intent, SecurityResourceLimits? limits = null)
    {
        ArgumentNullException.ThrowIfNull(intent);
        var request = Compile(intent);
        var contract = _contract ??
                       (_model ?? throw new InvalidOperationException("No semantic contract is configured.")).Freeze()
                       .CreateSnapshot();
        var graph = new SemanticRequestResolver(contract).Resolve(request);
        var operationGraph = SemanticOperationGraph.Create(SemanticOperationCompiler.Compile(graph));
        SemanticOperationGraphSafetyValidator.Validate(operationGraph, limits ?? new SecurityResourceLimits());
        return operationGraph;
    }

    /// <summary>Creates a contract-bound dynamic intent document.</summary>
    public SemanticIntentDocument CreateDocument(ReadIntent intent)
    {
        ArgumentNullException.ThrowIfNull(intent);
        return new SemanticIntentDocument(ContractFingerprint, intent).Validate();
    }

    /// <summary>Resolves a contract-bound intent document and rejects stale contracts before resolution.</summary>
    public SemanticIntentResolution ResolveDocument(SemanticIntentDocument document)
    {
        ArgumentNullException.ThrowIfNull(document);
        document.Validate();

        if (!string.Equals(document.ContractFingerprint, ContractFingerprint, StringComparison.Ordinal))
            throw new InvalidOperationException(
                $"Semantic intent document is bound to contract '{document.ContractFingerprint}', but the resolver uses '{ContractFingerprint}'.");

        var request = Compile(document.Intent);
        return new SemanticIntentResolution(document, request, ContractFingerprint);
    }

    /// <summary>Resolves a document directly to the canonical operation graph.</summary>
    public SemanticOperationGraph ResolveDocumentGraph(SemanticIntentDocument document)
    {
        var resolution = ResolveDocument(document);
        var contract = _contract ??
                       (_model ?? throw new InvalidOperationException("No semantic contract is configured.")).Freeze()
                       .CreateSnapshot();
        var graph = new SemanticRequestResolver(contract).Resolve(resolution.Request);
        return SemanticOperationGraph.Create(SemanticOperationCompiler.Compile(graph));
    }

    public string ContractFingerprint =>
        _contract?.ContractFingerprint ??
        (_model ?? throw new InvalidOperationException("No semantic contract is configured.")).Freeze()
        .ContractFingerprint;

    public SemanticRequest Compile(ReadIntent intent)
    {
        ArgumentNullException.ThrowIfNull(intent);

        var root = FindEntity(intent.RootEntity);
        var selections = intent.Selections.Select(selection =>
            CompileSelection(root, selection)).ToArray();

        var filter = intent.Filter is null ? null : CompileFilter(root, intent.Filter);
        var order = intent.Order?.Select(o => CompileOrder(root, o)).ToArray();

        return new SemanticRequest(
            root.Id,
            selections,
            new SemanticQueryOptions(filter, order, intent.Limit, intent.Offset, intent.After),
            intent.Security);
    }

    private SemanticSelection CompileSelection(SemanticEntity entity, ReadSelection selection)
    {
        if ((selection.Field is null) == (selection.Relationship is null))
            throw Invalid("A selection must specify exactly one field or relationship.");

        if (selection.Field is not null)
        {
            var field = FindField(entity, selection.Field);
            if (selection.EffectiveChildren.Count != 0)
                throw Invalid($"Field '{entity.Name}.{field.Name}' cannot have children.");
            return new SemanticSelection(field.Id, null, []);
        }

        if (selection.EffectiveChildren.Count == 0)
            throw Invalid($"Relationship '{entity.Name}.{selection.Relationship}' requires child selections.");

        var path = ResolveRelationshipPath(entity, selection.Relationship!);
        var compiledChildren = selection.EffectiveChildren.ToArray();

        // A logical traversal is deliberately expanded into its real semantic
        // relationship chain here. From this point onward the authorization
        // layer sees every hop, so tenant/field/relationship policies cannot be
        // bypassed by using a convenient alias such as Customer.transactions.
        IReadOnlyList<SemanticSelection> children = compiledChildren
            .Select(child => CompileSelection(GetEntity(path[^1].Target), child))
            .ToArray();

        for (var i = path.Count - 1; i >= 0; i--)
        {
            children = [new SemanticSelection(null, path[i].Id, children)];
        }

        return children[0];
    }

    private SemanticFilterExpression CompileFilter(SemanticEntity entity, ReadFilter filter) => filter switch
    {
        ReadFieldFilter field => new SemanticFieldFilter(
            FindField(entity, field.Field).Id,
            field.Operator,
            field.Value),

        ReadRelationshipFilter relationship => CompileRelationshipFilter(entity, relationship),

        ReadAndFilter andFilter when andFilter.Expressions.Count > 0 =>
            new SemanticAndFilter(andFilter.Expressions.Select(x => CompileFilter(entity, x)).ToArray()),

        ReadOrFilter orFilter when orFilter.Expressions.Count > 0 =>
            new SemanticOrFilter(orFilter.Expressions.Select(x => CompileFilter(entity, x)).ToArray()),

        ReadAndFilter => throw Invalid("AND filter cannot be empty."),
        ReadOrFilter => throw Invalid("OR filter cannot be empty."),
        _ => throw Invalid($"Unsupported read filter '{filter.GetType().Name}'.")
    };

    private SemanticOrderTerm CompileOrder(SemanticEntity root, ReadOrder order)
    {
        var entity = root;
        var path = new List<RelationshipId>();

        foreach (var relationshipName in order.EffectivePath)
        {
            var relationshipPath = ResolveRelationshipPath(entity, relationshipName);
            foreach (var relationship in relationshipPath)
            {
                path.Add(relationship.Id);
                entity = GetEntity(relationship.Target);
            }
        }

        var fieldId = order.Aggregate == SemanticOrderAggregate.Count
            ? entity.Identity.FieldId
            : FindField(entity, order.Field).Id;

        return new SemanticOrderTerm(
            fieldId,
            order.Direction,
            path,
            order.Aggregate);
    }

    private SemanticEntity FindEntity(string name)
    {
        if (_contract is not null)
            return _contract.TryResolveEntity(name, out var entity)
                ? entity
                : throw Invalid($"Unknown entity '{name}'.");

        return _model!.Entities.FirstOrDefault(x =>
                   string.Equals(x.Name, name, StringComparison.OrdinalIgnoreCase) ||
                   x.EffectiveAliases.Any(a => string.Equals(a.Name, name, StringComparison.OrdinalIgnoreCase)))
               ?? throw Invalid($"Unknown entity '{name}'.");
    }

    private SemanticEntity GetEntity(EntityId id) =>
        _contract?.Get(id) ?? _model!.Get(id);

    private static SemanticField FindField(SemanticEntity entity, string name)
    {
        var field = entity.Fields.FirstOrDefault(x =>
            string.Equals(x.Name, name, StringComparison.OrdinalIgnoreCase) ||
            x.EffectiveAliases.Any(a => string.Equals(a.Name, name, StringComparison.OrdinalIgnoreCase)));
        if (field is not null)
            return field;

        if (string.Equals(entity.Identity.Name, name, StringComparison.OrdinalIgnoreCase))
            return new SemanticField(entity.Identity.FieldId, entity.Identity.Name, typeof(object));

        throw Invalid($"Unknown field '{entity.Name}.{name}'.");
    }

    private IReadOnlyList<SemanticRelationship> ResolveRelationshipPath(SemanticEntity entity, string name)
    {
        var direct = entity.Relationships.FirstOrDefault(x =>
            string.Equals(x.Name, name, StringComparison.OrdinalIgnoreCase) ||
            x.EffectiveAliases.Any(a => string.Equals(a.Name, name, StringComparison.OrdinalIgnoreCase)));
        if (direct is not null)
            return [direct];

        SemanticTraversal? traversal = null;

        if (_contract is not null)
        {
            _contract.TryGetTraversal(entity.Id, name, out traversal);
        }
        else
        {
            _model!.TryGetTraversal(entity.Id, name, out traversal);
        }

        if (traversal is not null)
        {
            var current = entity;
            var result = new List<SemanticRelationship>(traversal.Path.Count);
            foreach (var relationshipId in traversal.Path)
            {
                var relationship = current.Relationships.FirstOrDefault(x => x.Id == relationshipId)
                                   ?? throw Invalid(
                                       $"Traversal '{name}' contains relationship '{relationshipId}' that is not defined on '{current.Name}'.");
                result.Add(relationship);
                current = GetEntity(relationship.Target);
            }

            return result;
        }

        throw Invalid($"Unknown relationship or semantic traversal '{entity.Name}.{name}'.");
    }

    private SemanticFilterExpression CompileRelationshipFilter(
        SemanticEntity entity,
        ReadRelationshipFilter filter)
    {
        var path = ResolveRelationshipPath(entity, filter.Relationship);
        if (path.Count > 1 && filter.Quantifier is not SemanticRelationshipQuantifier.Some)
        {
            throw Invalid(
                $"Logical traversal filter '{entity.Name}.{filter.Relationship}' currently supports only 'Some' quantification across multi-hop paths. " +
                "None/All require an explicit path-quantifier algebra so their meaning cannot be changed by expansion.");
        }

        SemanticFilterExpression predicate = CompileFilter(
            GetEntity(path[^1].Target),
            filter.Predicate);

        for (var i = path.Count - 1; i >= 0; i--)
            predicate = new SemanticRelationshipFilter(
                path[i].Id,
                i == path.Count - 1 ? filter.Quantifier : SemanticRelationshipQuantifier.Some,
                predicate);

        return predicate;
    }

    private static InvalidOperationException Invalid(string message) =>
        new($"Invalid read intent: {message}");
}